docs/merkletree/treetex/main.go

```diff
@@ -0,0 +1,216 @@
+// treetek is a command to produce LaTeX documents representing merkle trees.
+// Uses the Forest package.
+//
+// Usage: go run main.go | xelatex
+// This should generate a PDF file called treetek.pdf containing a drawing of
+// the tree.
+//
+package main
+
+import (
+	"flag"
+	"fmt"
+	"math/bits"
+	"strings"
+
+	"github.com/google/trillian/merkle"
+	"github.com/google/trillian/storage"
+)
+
+const (
+	preamble = `
+% Hash-tree
+% Author: treetek
+\documentclass[convert]{standalone}
+\usepackage[dvipsnames]{xcolor}
+\usepackage{forest}
+
+
+\begin{document}
+
+% Change colours here:
+\definecolor{inclusion}{rgb}{1,0.5,0.5}
+\definecolor{inclusion_ephemeral}{rgb}{1,0.7,0.7}
+\definecolor{perfect}{rgb}{1,0.9,0.5}
+\definecolor{target}{rgb}{0.5,0.5,0.9}
+
+\begin{forest}
+`
+
+	postfix = `\end{forest}
+\end{document}
+`
+	// maxLen is a suitably large maximum nodeID length for storage.NodeID.
+	maxLen = 64
+)
+
+var (
+	treeSize  = flag.Int64("tree_size", 23, "Size of tree to produce")
+	inclusion = flag.Int64("inclusion", -1, "Leaf index to show inclusion proof")
+
+	// nInfo holds nodeInfo data for the tree.
+	nInfo = make(map[string]nodeInfo)
+)
+
+// nodeInfo represents the style to be applied to a tree node.
+type nodeInfo struct {
+	incProof    bool
+	target      bool
+	perfectRoot bool
+	ephemeral   bool
+	leaf        bool
+}
+
+// String returns a string containing Forest attributes suitable for
+// rendering the node, given its type.
+func (n nodeInfo) String() string {
+	attr := make([]string, 0, 4)
+
+	// Figure out which colour to fill with:
+	fill := "white"
+	if n.perfectRoot {
+		attr = append(attr, "line width=2pt")
+	}
+	if n.incProof {
+		fill = "inclusion"
+		if n.ephemeral {
+			fill = "inclusion_ephemeral"
+			attr = append(attr, "draw, dotted")
+		}
+	}
+	if n.target {
+		fill = "target"
+	}
+	attr = append(attr, "fill="+fill)
+
+	if !n.ephemeral {
+		attr = append(attr, "draw")
+	}
+	if !n.leaf {
+		attr = append(attr, "circle, minimum size=3em")
+	} else {
+		attr = append(attr, "minimum size=1.5em")
+	}
+	return strings.Join(attr, ", ")
+}
+
+// setNodeInfo applies f to the nodeInfo associated with node k.
+func setNodeInfo(k string, f func(*nodeInfo)) {
+	n, ok := nInfo[k]
+	if !ok {
+		n = nodeInfo{}
+	}
+	f(&n)
+	nInfo[k] = n
+}
+
+/* perfect renders a perfect subtree.
```

Comment:
Why are some functions commented this way?

Reply:
Dusty brain, fixed.